File: android/src/main/java/me/lazerka/mf/android/JsonMapper.java
package me.lazerka.mf.android;

import android.util.Log;
import com.fasterxml.jackson.core.JsonParser;
import com.fasterxml.jackson.databind.DeserializationContext;
import com.fasterxml.jackson.databind.JsonDeserializer;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.deser.DeserializationProblemHandler;

import java.io.IOException;

/**
 * @author Dzmitry Lazerka
 */
public class JsonMapper extends ObjectMapper {
	private static final String TAG = JsonMapper.class.getName();

	public JsonMapper() {
		super();

		// Warn, but don't fail on unknown property.
		addHandler(new DeserializationProblemHandler() {
			@Override
			public boolean handleUnknownProperty(
					DeserializationContext deserializationContext,
					JsonParser jsonParser,
					JsonDeserializer<?> deserializer,
					Object beanOrClass,
					String propertyName
			) throws IOException {
				String msg = "Unknown property `" + propertyName + "` in " + beanOrClass;
				Log.w(TAG, msg);
				jsonParser.skipChildren();
				return true;
			}
		});
	}
}

File: android/src/main/kotlin/com/baraded/mf/io/JsonMapper.kt
/*
 *     Copyright (C) 2017 Dzmitry Lazerka
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU General Public License as published by
 *     the Free Software Foundation, either version 3 of the License, or
 *     (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU General Public License for more details.
 *
 *     You should have received a copy of the GNU General Public License
 *     along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

package com.baraded.mf.io

import com.baraded.mf.logging.LogService
import com.fasterxml.jackson.core.JsonParser
import com.fasterxml.jackson.databind.DeserializationContext
import com.fasterxml.jackson.databind.JsonDeserializer
import com.fasterxml.jackson.databind.deser.DeserializationProblemHandler

object JsonMapper : me.lazerka.mf.api.JsonMapper() {
    val log = LogService.getLogger(JsonMapper::class.java)

    init {
        // Warn, but don't fail on unknown property.
        addHandler(object : DeserializationProblemHandler() {
            override fun handleUnknownProperty(
                    deserializationContext: DeserializationContext?,
                    jsonParser: JsonParser?,
                    deserializer: JsonDeserializer<*>?,
                    beanOrClass: Any?,
                    propertyName: String?
            ): Boolean {
                val msg = "Unknown property `$propertyName` in $beanOrClass"
                log.warn(msg)
                jsonParser?.skipChildren()
                return true
            }
        })

    }

}